packages/sdk/src/container/teamsfx.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { satisfies } from "semver";
+import { Component, ComponentApiNames, ComponentContainer, Mapping, TeamsFx } from "./types";
+import { Logger, InternalLogger, LogLevel, LogFunction } from "../util/logger";
+
+export class TeamsFxContainer<T extends Component> implements ComponentContainer, TeamsFx {
+  private registry: Map<string, Component>;
+  private initialized: Map<string, boolean>;
+  private loggers: Map<string, InternalLogger>;
+
+  private logLevel?: LogLevel;
+
+  constructor(components: T[]) {
+    this.registry = new Map<string, Component>();
+    this.initialized = new Map<string, boolean>();
+    this.loggers = new Map<string, InternalLogger>();
+
+    for (const component of components) {
+      this.addComponent(component);
+    }
+  }
+
+  private addComponent(component: T): void {
+    this.registry.set(component.name, component);
+    const logger = new InternalLogger(component.name, this.logLevel);
+    this.loggers.set(component.name, logger);
+
+    const functionKeys = Object.getOwnPropertyNames(Object.getPrototypeOf(component));
+    functionKeys.forEach((key: string) => {
+      const method = component[key];
+      if (method instanceof Function && !ComponentApiNames.includes(key) && !key.startsWith("_")) {
+        (this as Mapping)[key] = (...args: any) => {
```

Comment:
2 same method names will be conflict and should be avoided.
The private methods will be cloned too but users cannot call them because they are not included in type definition.